<a href="https://colab.research.google.com/github/AlexBB999/Thinkful/blob/master/32_3_Dask.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**ASSIGNMENT ANSWERS AT BOTTOM**

##**Assignments**

In this assignment, you'll continue working with the Credit Card Fraud Detection dataset from Kaggle.

So, you should load the dataset before working on the following tasks.

To complete this assignment, create a Jupyter notebook containing your solutions to the following tasks and submit as a link on Github.

How many transactions do we have in total?

How many transactions are fraud and how many are not?

What is the maximum amount in fraud transactions?

Submit your work below, and plan on discussing it with your mentor.

 You can also take a look at this example solution.

In [0]:
#!pip install --upgrade "dask[complete]"

In [16]:
pip install -U ipykernel

     |████████████████████████████████| 122kB 2.8MB/s 
ERROR: google-colab 1.0.0 has requirement ipykernel~=4.10, but you'll have ipykernel 5.2.1 which is incompatible.
  Found existing installation: ipykernel 4.10.1
    Uninstalling ipykernel-4.10.1:
      Successfully uninstalled ipykernel-4.10.1


In [1]:
import warnings
warnings.filterwarnings("ignore")

from dask.distributed import Client, progress

client = Client(n_workers=4, threads_per_worker=2, memory_limit='2GB')
client

<Client: 'tcp://127.0.0.1:44333' processes=4 threads=8, memory=8.00 GB>

In [0]:
import matplotlib.pyplot as plt

In [0]:
# Dataframes implement the Pandas API
import dask.dataframe as dd

# This loads the data into Dask dataframe
df = dd.read_csv('https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/creditcard.csv')

In [3]:
print(df)

Dask DataFrame Structure:
                Time       V1       V2       V3       V4       V5       V6       V7       V8       V9      V10      V11      V12      V13      V14      V15      V16      V17      V18      V19      V20      V21      V22      V23      V24      V25      V26      V27      V28   Amount  Class
npartitions=3                                                                                                                                                                                                                                                                                   
               int64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  int64
                 ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      

In [4]:
df.dtypes

Time        int64
V1        float64
V2        float64
V3        float64
V4        float64
V5        float64
V6        float64
V7        float64
V8        float64
V9        float64
V10       float64
V11       float64
V12       float64
V13       float64
V14       float64
V15       float64
V16       float64
V17       float64
V18       float64
V19       float64
V20       float64
V21       float64
V22       float64
V23       float64
V24       float64
V25       float64
V26       float64
V27       float64
V28       float64
Amount    float64
Class       int64
dtype: object

In [5]:
df.head()

Time        V1        V2        V3  ...       V27       V28  Amount  Class
0     0 -1.359807 -0.072781  2.536347  ...  0.133558 -0.021053  149.62      0
1     0  1.191857  0.266151  0.166480  ... -0.008983  0.014724    2.69      0
2     1 -1.358354 -1.340163  1.773209  ... -0.055353 -0.059752  378.66      0
3     1 -0.966272 -0.185226  1.792993  ...  0.062723  0.061458  123.50      0
4     2 -1.158233  0.877737  1.548718  ...  0.219422  0.215153   69.99      0

[5 rows x 31 columns]

In [6]:
df.describe()

Dask DataFrame Structure:
                  Time       V1       V2       V3       V4       V5       V6       V7       V8       V9      V10      V11      V12      V13      V14      V15      V16      V17      V18      V19      V20      V21      V22      V23      V24      V25      V26      V27      V28   Amount    Class
npartitions=1                                                                                                                                                                                                                                                                                       
               float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64
                   ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...
Dask Name: describe-numeric, 350 tasks

In [7]:
df.describe().compute()

ValueError: ignored

Oops! We got an error. That's because Dask tried to infer the data types of the columns it loads.

 However, Dask doesn't consider all the values in each column when deciding the data types due to the lazy evaluation.
 
**Instead, it randomly sample a subset of the columns and guess what their data types are. **
  
In case we encounter these type of situations, we need to specify the column types manually.

 Moreover, **setting the assume_missing parameter of the read_csv() function to True** 

would also be helpful in some situations as it informs Dask that all integer columns that aren’t specified in dtype parameter are assumed to contain missing values, and are converted to floats.

Now, let's read our dataset again but specify the type of the Time columns as float64:

In [0]:
df = dd.read_csv('https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/creditcard.csv', dtype={'Time': 'float64'})

In [9]:
df.describe().compute()

Time            V1  ...         Amount          Class
count  284807.000000  2.848070e+05  ...  284807.000000  284807.000000
mean    94813.859575  6.003539e-16  ...      88.349619       0.001727
std     47488.145955  1.958696e+00  ...     250.120109       0.041527
min         0.000000 -5.640751e+01  ...       0.000000       0.000000
25%     49346.000000 -7.868600e-01  ...       6.840000       0.000000
50%     94872.000000  9.301532e-02  ...      24.990000       0.000000
75%    165830.000000  1.984826e+00  ...      84.910000       0.000000
max    172792.000000  2.454930e+00  ...   25691.160000       1.000000

[8 rows x 31 columns]

In [10]:
df2 = df[df["Amount"] > 10000]
df2

Dask DataFrame Structure:
                  Time       V1       V2       V3       V4       V5       V6       V7       V8       V9      V10      V11      V12      V13      V14      V15      V16      V17      V18      V19      V20      V21      V22      V23      V24      V25      V26      V27      V28   Amount  Class
npartitions=3                                                                                                                                                                                                                                                                                     
               float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  float64  int64
                   ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...    ...
                   ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...    ...
                   ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...      ...    ...
Dask Name: getitem, 18 tasks

In [12]:
df[df["Amount"] > 10000].compute()

Time         V1         V2  ...        V28    Amount  Class
46841   42951.0 -23.712839 -42.172688  ...   2.253662  12910.93      0
54018   46253.0 -21.780665 -38.305310  ...   2.082488  11898.09      0
58465   48401.0 -36.802320 -63.344698  ...   3.450224  19656.53      0
30325   95286.0 -34.549296 -60.464618  ...   3.250802  18910.00      0
48486  119713.0 -20.924897 -37.943452  ...   2.001492  11789.84      0
32983  166198.0 -35.548539 -31.850484  ... -15.430084  25691.16      0
42461  172273.0  -9.030538 -11.112584  ...  -2.713136  10199.44      0

[7 rows x 31 columns]

In [14]:
df.loc[df["Amount"] > 10000,:].compute()

Time         V1         V2  ...        V28    Amount  Class
46841   42951.0 -23.712839 -42.172688  ...   2.253662  12910.93      0
54018   46253.0 -21.780665 -38.305310  ...   2.082488  11898.09      0
58465   48401.0 -36.802320 -63.344698  ...   3.450224  19656.53      0
30325   95286.0 -34.549296 -60.464618  ...   3.250802  18910.00      0
48486  119713.0 -20.924897 -37.943452  ...   2.001492  11789.84      0
32983  166198.0 -35.548539 -31.850484  ... -15.430084  25691.16      0
42461  172273.0  -9.030538 -11.112584  ...  -2.713136  10199.44      0

[7 rows x 31 columns]

**GROUPBY**

In [15]:
df2 = df.groupby("Class")["Amount"].mean().compute()
df2

Class
0     88.291022
1    122.211321
Name: Amount, dtype: float64

##**Using apply() function**

**.apply() function operates row by row and each operation is independent to each other**

In [16]:
df["Amount2"] = df["Amount"].apply(lambda x: 1 if x>10000 else 0, meta=('Amount2', 'int64'))
df["Amount2"].head(20)

0     0
1     0
2     0
3     0
4     0
5     0
6     0
7     0
8     0
9     0
10    0
11    0
12    0
13    0
14    0
15    0
16    0
17    0
18    0
19    0
Name: Amount2, dtype: int64

In [0]:
df = df.persist()

In [0]:
client.close()

**////////////////////////////////////////////////////////////////////////////////////**

**#1 How many transactions do we have in total?**

In [19]:
https://drive.google.com/open?id=1eB0KlE2rB3mPwVtWWXj2kFevYLynhd3x

SyntaxError: ignored

In [0]:
df = dd.read_csv('/content/creditcard.csv', assume_missing=True)

In [34]:
df.head()

Time        V1        V2        V3  ...       V27       V28  Amount  Class
0   0.0 -1.359807 -0.072781  2.536347  ...  0.133558 -0.021053  149.62    0.0
1   0.0  1.191857  0.266151  0.166480  ... -0.008983  0.014724    2.69    0.0
2   1.0 -1.358354 -1.340163  1.773209  ... -0.055353 -0.059752  378.66    0.0
3   1.0 -0.966272 -0.185226  1.792993  ...  0.062723  0.061458  123.50    0.0
4   2.0 -1.158233  0.877737  1.548718  ...  0.219422  0.215153   69.99    0.0

[5 rows x 31 columns]

In [36]:
df.tail()

Time         V1         V2  ...       V28  Amount  Class
43014  172786.0 -11.881118  10.071785  ...  0.823731    0.77    0.0
43015  172787.0  -0.732789  -0.055080  ... -0.053527   24.79    0.0
43016  172788.0   1.919565  -0.301254  ... -0.026561   67.88    0.0
43017  172788.0  -0.240440   0.530483  ...  0.104533   10.00    0.0
43018  172792.0  -0.533413  -0.189733  ...  0.013649  217.00    0.0

[5 rows x 31 columns]

In [40]:
df['Class'].nunique().compute()

2

In [35]:
df.describe().compute()

Time            V1  ...         Amount          Class
count  284807.000000  2.848070e+05  ...  284807.000000  284807.000000
mean    94813.859575  6.003539e-16  ...      88.349619       0.001727
std     47488.145955  1.958696e+00  ...     250.120109       0.041527
min         0.000000 -5.640751e+01  ...       0.000000       0.000000
25%     49346.000000 -7.868600e-01  ...       6.840000       0.000000
50%     94872.000000  9.301532e-02  ...      24.990000       0.000000
75%    165830.000000  1.984826e+00  ...      84.910000       0.000000
max    172792.000000  2.454930e+00  ...   25691.160000       1.000000

[8 rows x 31 columns]

284,807 transactions

**#2 How many transactions are fraud and how many are not?**

In [63]:
df['Class'].value_counts().compute()

0.0    284315
1.0       492
Name: Class, dtype: int64

**492 Fraud    -- 284315  Good**

**What is the maximum amount in fraud transactions?  <---POORLY WORDED!**

**AMOUNT USUALLY MEANS SUM -- SHOULD SAY WHAT IS MAXIMUM VALUE**

In [66]:
df[df['Class']==1]['Amount'].sum().compute()

60127.97

In [67]:
df[df['Class']==1]['Amount'].max().compute()

2125.87